### PR TITLE
Sort operator for read APIs

### LINKED ISSUE #223: Sort operator for read APIs
Server-side support already exists for this

### PR DESCRIPTION
**Pull request by Shubham8287**
This PR closed #223
/claim #223

### CODE REVIEW

File: driver/types.go [hunk 1]
```diff
@@ -61,12 +61,25 @@ type (
 	Vector json.RawMessage
 )
 
+// ReadOptions modifies read request behavior.
+type ReadOptions struct {
```
Comment (Shubham8287) [inline driver/types.go:65]: At first, I created `ReadOption` as Message type in api.proto but realised, it is not needed there and can be split between `ReadRequestOptions` and `Sort`. So just defining it as struct here. Let me know, if this is not fine.

**Comment by Shubham8287** [comment]
Hi @efirs, I am not sure, what practices do we follow here(as there is no Contribution.md) here. Incase, we need to add tests for Sort. Please let me know, and some pointers for where do add tests would also be helpful.

**Comment by Shubham8287** [comment]
It looks Checks do not triggers automatically, How to trigger them manually? `make tests` was Green on local for me.

File: driver/http.go [hunk 2]
```diff
@@ -669,12 +669,18 @@ func (c *httpCRUD) deleteWithOptions(ctx context.Context, collection string, fil
 
 func (c *httpCRUD) readWithOptions(ctx context.Context, collection string, filter Filter, fields Projection, options *ReadOptions) (Iterator, error) {
 	ctx = setHTTPTxCtx(ctx, c.txCtx, c.cookies)
-
+	var sort []json.RawMessage
+	if options.Sort != nil {
+		sort = []json.RawMessage{json.RawMessage(options.Sort)}
```
Comment (efirs) [inline driver/http.go:674]: I think we need to fix the HTTP client to accept just json.RawMessage.
To do that you can change `yq_fix_json ReadRequest sort.items`
to `yq_fix_json ReadRequest sort` in scripts/fix_openapi.sh
and run `make generate`.

**Comment by efirs** [comment]
> Hi @efirs, I am not sure, what practices do we follow here(as there is no Contribution.md to follow). Incase, we need to add tests for Sort. Please let me know, and some pointers for where do add tests will also be helpful.

Yes, please add tests.
For the high level you can modify `read_with_options_filter_all` test in collection_test.go
For the low level you can modify `read_with_collation` in driver_test.go

gomocks may be tricky at times let me know if you face any issues.

**Comment by efirs** [comment]
> It looks Checks do not triggers automatically, How to trigger them manually? `make tests` was Green on local for me.

After you update PR next time will be able to start the tests.

**Comment by Shubham8287** [comment]
Thanks for reviewing @efirs :)

**Comment by efirs** [comment]
@Shubham8287 I suggest you to reset your changes to api/client/v1/api/http.go and rebase first. Then regenerate.
This would resolve the conflict.

**Comment by Shubham8287** [comment]
@efirs After regenerating, I am not sure how there are changes in api.proto in submodule. Please let me know,  if that's not intended. Sorry I am working with Grpc for the first time so not sure about this.

**Comment by efirs** [comment]
Assuming that you have all your changes commit let me give instructions how to clean it up:
```
git checkout -b tmp_save_for_sort #backup your branch
git checkout - #go back to the original branch
git reset main # reset you commits upto main
git checkout api #clean the api changes
git add -u
git commit -m "feat: Sorting support"
git checkout main 
git pull origin main #pull latest changes from github
git checkout - #go back the original branch
git rebase main # reabase your changes on top of main
make generate #only http.go should be modified
git add -u
git commit --amend
git push -f origin {your branch name}
```

**Comment by Shubham8287** [comment]
@efirs , Thanks for the guidance, I have also added Tests. yesterday, I messed up a little bit by editing the submodule but fix those changes. Tests are again passing for me. Please review.
```
ubuntu@DESKTOP-C07UJHE:~/tigris-client-go$ make test
mkdir -p mock/api
mockgen -package mock -destination mock/driver.go github.com/tigrisdata/tigris-client-go/driver \
        Driver,Tx,Database,Iterator,SearchResultIterator,SearchClient,SearchIndexResultIterator
mockgen -package api -destination mock/api/grpc.go github.com/tigrisdata/tigris-client-go/api/server/v1 \
        TigrisServer,AuthServer,ManagementServer,ObservabilityServer,SearchServer
go mod download
go test -cover -race -tags=test  ./...
?       github.com/tigrisdata/tigris-client-go/api      [no test files]
?       github.com/tigrisdata/tigris-client-go/api/client/v1/api        [no test files]
?       github.com/tigrisdata/tigris-client-go/api/server/v1    [no test files]
?       github.com/tigrisdata/tigris-client-go/code     [no test files]
?       github.com/tigrisdata/tigris-client-go/config   [no test files]
ok      github.com/tigrisdata/tigris-client-go/driver   3.487s  coverage: 84.3% of statements
ok      github.com/tigrisdata/tigris-client-go/fields   (cached)        coverage: 96.6% of statements
ok      github.com/tigrisdata/tigris-client-go/filter   (cached)        coverage: 98.2% of statements
?       github.com/tigrisdata/tigris-client-go/mock     [no test files]
?       github.com/tigrisdata/tigris-client-go/mock/api [no test files]
ok      github.com/tigrisdata/tigris-client-go/schema   (cached)        coverage: 93.8% of statements
ok      github.com/tigrisdata/tigris-client-go/search   0.281s  coverage: 91.1% of statements
ok      github.com/tigrisdata/tigris-client-go/sort     (cached)        coverage: 93.8% of statements
?       github.com/tigrisdata/tigris-client-go/test     [no test files]
ok      github.com/tigrisdata/tigris-client-go/tigris   0.211s  coverage: 84.5% of statements
ok      github.com/tigrisdata/tigris-client-go/timeseries       0.064s  coverage: 91.7% of statements
```

**Review by efirs — APPROVED** [review]

**Comment by efirs** [comment]
The diff looks very cool, @Shubham8287 !
Thanks for contributing!

**Comment by efirs** [comment]
One more thing before I merge,
Could you squash the commits and sign it?
You can sign by running `git commit --amend --gpg-sign`.

**Comment by efirs** [comment]
Nevermind, I merged it.
Thanks again, for the contribution!

**Comment by Shubham8287** [comment]
Thanks @efirs for guiding me with patience! This is a new tech stack for me, but the code looks really well maintained and now I also got a glimpse of it. Looking forward to contributing more :)

**Pull request merged**